From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: purpurcof <purpurcof@users.noreply.github.com>
Date: Mon, 27 Jul 2026 12:00:00 +0300
Subject: [PATCH] Move ClientboundCommandsPacket creation off the main thread

Build the packet on the async command sending pool to avoid
enumerateNodes()/createEntries() lag on login with many commands.

diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index de5b9c69ee6579c963bb00a3f2628247af9951e8..11223344556677889900aabbccddeeff11223344 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -524,5 +524,9 @@ public class Commands {
         }
         // CraftBukkit end
         } // Gale - Purpur - skip PlayerCommandSendEvent if there are no listeners
-        player.connection.send(new ClientboundCommandsPacket(rootCommandNode, COMMAND_NODE_INSPECTOR));
+        final net.minecraft.server.MinecraftServer server = net.minecraft.server.MinecraftServer.getServer();
+        final net.minecraft.server.network.ServerGamePacketListenerImpl connection = player.connection;
+        COMMAND_SENDING_POOL.execute(() -> {
+            ClientboundCommandsPacket packet = new ClientboundCommandsPacket(rootCommandNode, COMMAND_NODE_INSPECTOR);
+            if (!server.isStopped()) {
+                server.execute(() -> connection.send(packet));
+            }
+        });
     }
